doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,653 @@
+
+- **Can Tick?**: Whether this function is allowed to tick the QUIC state
+  machine and potentially perform network I/O.
+
+- **CSHL:** Connection/Stream/Handshake Layer classification.
+  This can be one of:
+
+    - **HL:** This is a handshake layer related call. It should be supported
+      on a QUIC connection SSL object, forwarding to the handshake layer
+      SSL object.
+
+      Whether we allow QUIC stream SSL objects to have these calls forwarded is
+      TBD.
+
+    - **HL-Forbidden:** This is a handshake layer related call, but it is
+      inapplicable to QUIC, so it is not supported.
+
+    - **C:** Not handshake-layer related. QUIC connection SSL object usage only.
+      Fails on a QUIC stream SSL object.
+
+    - **CS:** Not handshake-layer related. Can be used on any QUIC SSL object.
+
+### Existing APIs
+
+#### `SSL_set_connect_state`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Never         | No        | HL            |
+
+#### `SSL_set_accept_state`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Never         | No        | HL            |
+
+**Note:** Attempting to proceed in this state will not function for now because
+we do not implement server support at this time. However, the semantics of this
+function as such are unchanged.
+
+#### `SSL_is_server`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Never         | No        | HL            |
+
+#### `SSL_connect`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | HL            |
+
+Simple composition of `SSL_set_connect_state` and `SSL_do_handshake`.
+
+#### `SSL_accept`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | HL            |
+
+Simple composition of `SSL_set_accept_state` and `SSL_do_handshake`.
+
+#### `SSL_do_handshake`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | HL            |
+
+**Note:** Idempotent if handshake already completed.
+
+**Blocking Considerations:** Blocks until handshake completed if in blocking
+mode.
+
+#### `SSL_read`, `SSL_read_ex`, `SSL_peek`, `SSL_peek_ex`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | CS            |
+
+**Blocking Considerations:** Blocks until at least one byte is available or an
+error occurs if in blocking mode (including the peek functions).
```

Comment:
I'm not sure about "Can tick" here (and for SSL_write below).

In TLS calling "SSL_read" may indicate that the application believes that the underlying socket is readable. If we get an "SSL_read" call then we assume that performing a read on the socket is ok.

In QUIC it is a lot less clear whether an application calling SSL_read on a stream object is also indicating readability of the underlying socket.

Reply:
Unclear to me how this could work any other way. There can be data buffered internally or new data available to read from the network and this is opaque to the application. This is reflected in the fact that `SSL_ERROR_WANT_READ/WRITE`, of necessity, no longer has anything to do with network-level I/O readiness.